wallet/dapps-communication.md

```diff
@@ -1,11 +1,696 @@
+
+<table>
+<thead>
+<tr>
+<th>Direction</th>
+<th>Type</th>
+<th>Format</th>
+<th>Example</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>Webpage to extension</td>
+<td><code>account.list</code></td>
+<td>none</td>
+<td><code>null</code></td>
+</tr>
+<tr>
+<td>Extension to webpage</td>
+<td><code>account.list.response</code></td>
+<td>
+
+```yaml
+type: array
+items:
+  type: object
+  properties:
+    address:
+      type: string
+    name:
+      type: string
+  required:
+    - address
+```
+
+</td><td>
+  
+```json
+[
+  {
+    "address": "A12...",
+    "name": "Account 1"
+  },
+  {
+    "address": "A13...",
+    "name": "Account 2"
+  },
+]
+```
+
+</td></tr></tbody></table>
+
+##### Balance
+
+<table>
+<thead>
+<tr>
+<th>Direction</th>
+<th>Type</th>
+<th>Format</th>
+<th>Example</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>Webpage to extension</td>
+<td><code>account.balance</code></td>
+<td>
+
+```yaml
+type: object
+properties:
+  address:
+  type: string
+required:
+  - address
+```
+
+</td><td>
+
+```json
+{
+  "address": "A12..."
+}
+```
+
+</td></tr><tr>
+<td>Extension to webpage</td>
+<td><code>account.balance.response</code></td>
+<td>
+
+```yaml
+type: object
+properties:
+  balance:
+    type: string
+    format: BigInt
+required:
+  - balance
+```
+
+</td><td>
+  
+```json
+{
+  "balance": "1000"
+}
+```
+
+</td></tr></tbody></table>
+
+##### Delete
+
+<table>
+<thead>
+<tr>
+<th>Direction</th>
+<th>Type</th>
+<th>Format</th>
+<th>Example</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>Webpage to extension</td>
+<td><code>account.delete</code></td>
+<td>
+
+```yaml
+type: object
+properties:
+  address:
+  type: string
+required:
+  - address
+```
+
+</td>
+<td>
+
+```json
+{
+  "address": "A12..."
+}
+```
+
+</td>
+</tr>
+<tr>
+<td>Extension to webpage</td>
+<td><code>account.delete.response</code></td>
+<td>
+
+```yaml
+type: object
+properties:
+  response:
+    type: string
+    enum: ["OK", "REFUSED", "ERROR"]
+  error:
+    type: string
+required:
+  - response
+```
+
+</td><td>
+  
+```json
+{
+  "response": "REFUSED"
+}
+```
+
+</td></tr></tbody></table>
+
+##### Import
+
+<table>
+<thead>
+<tr>
+<th>Direction</th>
+<th>Type</th>
+<th>Format</th>
+<th>Example</th>
+</tr>
+</thead>
+<tbody>
+<tr>
+<td>Webpage to extension</td>
+<td><code>account.import</code></td>
+<td>
+
+```yaml
+type: object
+properties:
+  privateKey:
+    type: string
+  publicKey:
+    type: string
+required:
+  - privateKey
+  - publicKey
+```
+
+</td><td>
+
+```json
+{
+  "privateKey": "S12...",
+  "publicKey": "P12...",
+}
+```
+
+</td>
+</tr>
+<tr>
+<td>Extension to webpage</td>
+<td><code>account.import.response</code></td>
+<td>
+
+```yaml
+type: object
+properties:
+  response:
+    type: string
+    enum: ["OK", "REFUSED", "ERROR"]
+  error:
+    type: string
+required:
+  - response
+```
+
+</td><td>
+  
+```json
+{
+  "response": "ERROR",
+  "error": "No connection with blockchain"
+}
```

Comment:
Non blocking, of course: isn't better use **message** as key instead of **error** since **REFUSED** could also be followed by a message to inform the reason was refused!? 🤔  (not sure if is applicable).

```suggestion
{
  "response": "ERROR",
  "message": "No connection with blockchain"
}
```

Reply:
Non blocking: would we possible run in a situation where we want to inform DEVs (on browser console) about something deprecated!? A new field missing, a new way to sign, a new version, etc!?

If yes, shouldn't we answer back using the well know HTTP standards? 200 (ok), 403 (forbidden), 426 (upgrade require) , etc? 🤔

Reply:
> Non blocking, of course: isn't better use **message** as key instead of **error** since **REFUSED** could also be followed by a message to inform the reason was refused!? thinking (not sure if is applicable).

Cool, we should do that

Reply:
> Non blocking: would we possible run in a situation where we want to inform DEVs (on browser console) about something deprecated!? A new field missing, a new way to sign, a new version, etc!?
> 
> If yes, shouldn't we answer back using the well know HTTP standards? 200 (ok), 403 (forbidden), 426 (upgrade require) , etc? thinking

Don't you think that the message is sufficient and easier to understand than HTTP response code for a human (I don't believe that this kind of returned will be automaticaly processed by the code...) ?

Reply:
Indeed, it could be overkill. I don't disagree... I was just considering that potentially we could have more options than the 3 ones: **ok**, **refused**, **error**. If we adopt the well know HTTP palette we have intrinsic all kind of standard possibilities (if needed, of course). 

However, in the future, by some reason, we decide to include a new status (WARNING) for example, people need to be informed about the new status to eventually rely on that answer. Instead, using HTTP they are already aware that almost all range of API answer are covered.

But, please, feel free to go with you have defined. It's completely acceptable and simple.

Reply:
I have renamed error into message.

I think the enum is enough and http status code is too much.

Reply:
Thank you! 

Sure, no worries for the status codes.